### DISCUSSION TITLE
Startup arrangements

### DISCUSSION THREAD

**Discussion by devmatteini** [wezterm/wezterm#3236] [Q&A] [answered]
Hi, and thanks for this amazing terminal emulator!

I was looking for a way to set up terminal tabs and panes (and shell commands to run) for specific projects.

I've tried two approaches: 

### 1. Bash scripts

After opening a wezterm terminal, I run `/path/to/script.sh` and my terminal tabs and panes are created.

```bash
#!/usr/bin/env bash

set -eo pipefail

PROJECT_DIR="$HOME/dev"

run_command() {
    local pane_id
    local command_to_run
    pane_id="$1"
    command_to_run="$2"

    # https://github.com/wez/wezterm/discussions/2202#discussioncomment-3052321
    echo -e "$command_to_run\n" | wezterm cli send-text --no-paste --pane-id "$pane_id"
}

# Creates panes
main_pane=$(wezterm cli spawn --cwd "$PROJECT_DIR")
second_pane=$(wezterm cli split-pane --right --pane-id "$main_pane")

# Starts commands
run_command "$main_pane" 'echo "this is 1st pane"'
run_command "$second_pane" 'echo "this is 2nd pane"'
```

###  2. Lua and `gui-startup`

```lua
local wezterm = require "wezterm"
local mux = wezterm.mux

wezterm.on('gui-startup', function(cmd)
  local project_dir = wezterm.home_dir .. '/dev'
  local tab, main_pane, window = mux.spawn_window {
    workspace = 'test',
    cwd = project_dir,
  }
  local second_pane = main_pane:split {
    direction = 'Right',
    size = 0.5,
    cwd = project_dir,
  }

  main_pane:send_text 'echo "this is 1st pane"\n'
  second_pane:send_text 'echo "this is 2st pane"\n'
end)

return {}
```

## Questions

The `gui-startup` it's definitely easier to use, but if you define multiple workspaces for different projects once you open a terminal they all get created (and each command like `main_pane:send_text` is executed).
So for now I'm using the bash scripts to prevent this issue of `gui-startup` and workspaces.

Maybe I'm missing something or not doing things properly with the lua config, so feel free to suggest any improvements.

I would really like to able to do something like this:

```bash
wezterm --startup /path/to/file.lua
```

where `/path/to/file.lua` is something similar to what is possible to do with `gui-startup` in `wezterm.lua`

**Comment by wez** [comment, marked as answer]
You might want to consider doing something like:

```
MY_STARTUP_MODE=foo wezterm
```

then your startup event handler might use that env var to change its behavior:

```lua
wezterm.on('gui-startup', function(cmd)
   local mode = os.getenv("MY_STARTUP_MODE")
   if mode == "foo" then
      -- do something
   end
end)
```

Another option is:

```
wezterm --config-file /path/to/file.lua
```

but I think that would result in more complex logic in your config file(s).

**Reply by devmatteini** [reply]
I'll try the `MY_STARTUP_MODE=foo wezterm` and see how it goes.

I guess if I use `wezterm --config-file /path/to/file.lua` I should probably split my config files to have common configs:

`base.lua`
```lua
local wezterm = require "wezterm"
return {
  font_size = 15.0,
  -- other configs...
}
```

`/path/to/file.lua`
```lua
local wezterm = require "wezterm"
local base = require "base"

wezterm.on('gui-startup', function(cmd)
   -- arrange tabs & panes
end)

return base
```

Not tested this, but it could actually be pretty nice!

**Reply by devmatteini** [reply]
At the end, I used the environment variable approach.
I wrote a blog post with my implementation if it could be helpful to someone: https://cosimomatteini.com/blog/terminal-configurations-with-wezterm

**Reply by wez** [reply]
For that file_exists checking stuff, I think you may be able to simplify it a bit to avoid guessing about the path.

```lua
-- pcall allows catching an error
local status, project_module = pcall(function()
   -- lua will automatically convert `projects.foo` into `/home/USER/.config/wezterm/projects/foo.lua`
   -- when processing this require statement
   return require("projects." .. os.getenv("WZ_PROJECT"))
end)
if not status then
   -- in the failure case, project_module is the error message
   wezterm.log_error("unable to import " .. project_module
   return
end
-- otherwise, `project_module` is the imported module
```

**Reply by devmatteini** [reply]
Nice, thanks for the suggestion!

**Comment by Issafalcon** [comment]
Hi guys!

I'm new to WezTerm, but I can already tell it's going to be an ideal tool for me to create my perfect workstation setup (Windows with WSL2 and the odd remote linux session).

Thanks for this project @wez , and thanks for putting together the guide post for setting up project specific layouts @devmatteini ! I'm setting this up as we speak!

On a related note, I've been using tmux quite a bit for my multiplexing in WSL, and I'm liking the ability to manage the sessions, and seamlessly switch between them in the same terminal window. I'm planning on getting multiplexing for my Windows workspaces in Wezterm to create a similar sort of setup. Playing around with Wezterm Workspaces and reading this discussion made me think of a couple of things:

1. Is it possible to hook into the existing selection list functionality in Wezterm, and display a custom list? 
2. Enabling this, could one then have a predefined set of 'project' files that could each just run a lua function to manually spin up a new Workspace and switch to it whilst keeping the existing Workspaces alive?
3. Will unix connection over to WSL2 ever be possible so I can ditch tmux and just use WezTerm for everything instead? (this is probably one for the guys over at Microsoft)

I like the concept of being able to spin up new Wezterm windows, each with their own project layouts (which I think is what your guide allows @devmatteini ?) but it would be great if you could take it that one step further and switch to a new project using a simple keyboard shortcut and selection, in the same Wezterm window.

Let me know if I've completely missed anything about Wezterm functionality here. I may have done! 😄

**Reply by wez** [reply]
> 1. Is it possible to hook into the existing selection list functionality in Wezterm, and display a custom list?

Not currently, but I've been thinking about this recently.

> 3. Will unix connection over to WSL2 ever be possible so I can ditch tmux and just use WezTerm for everything instead? (this is probably one for the guys over at Microsoft)

A consequence of the WSL 2 architecture change from WSL 1 is that direct unix socket bridging to the host is not possible.
If MS change that, then we can do something with it, but otherwise: you'll need to treat the WSL 2 container similarly to a remote unix host and use either SSH or TLS multiplexing to reach inside it.

**Reply by Issafalcon** [reply]
Nice!

Thanks for the rapid response. I'm glad it's on your radar. It would a be a very cool feature to add I think!

I'll stick to tmux on WSL2 for now, and keep and eye out for MS to get it sorted. Knowing MS turnaround time, I'll probably know some Rust by then, and hopefully could contribute!

**Reply by devmatteini** [reply]
> I like the concept of being able to spin up new Wezterm windows, each with their own project layouts (which I think is what your guide allows @devmatteini ?)

Yes, my guide was exactly for that!

> but it would be great if you could take it that one step further and switch to a new project using a simple keyboard shortcut and selection, in the same Wezterm window.

Yeah, this would be awesome to have! Surely much easier than my setup with environment variable.
 @wez do you think this could be something useful to have implemented in WezTerm?

I'm willing to help if this is something you might consider doing!

**Reply by wez** [reply]
I added [InputSelector](https://wezfurlong.org/wezterm/config/lua/keyassignment/InputSelector.html) which should show up in the nightly downloads within the next 20-30 minutes.

**Reply by Issafalcon** [reply]
Legend!

**Reply by Issafalcon** [reply]
I've taken some ideas from your guide @devmatteini and had a rough go at setting up a custom projects list for the InputSelector.

https://github.com/Issafalcon/wezterm-config

Feel free to use it as an example / improve on it for your own needs.

**Comment by ciyer** [comment]
I just wanted to thank everyone in this thread for their contributions (and particularly wez for the code changes). 

I have been using [iTerm2](https://iterm2.com) for around 15 years (yikes!), and I have some scripting set up there to realize this same feature: environment set up for my different projects. iTerm2 is a great terminal emulator, but I do look at alternatives every now and then, and I checked out WezTerm a couple of years ago. I really liked what I saw, but I rely my set up scripts to be productive, and I could not figure out a way to do this in WezTerm at that time.

Fast-forward to today, thanks to some commits that have happened in the intervening time, and this thread which explains how to take advantage of them, I am now able to replicate my per-project window/environment set ups in WezTerm, and the Lua-based scripting environment is truly enjoyable to work in.

Thanks again!
